v2/apiserver/internal/core/workers.go

```diff
@@ -390,6 +402,85 @@ func (w *workersService) UpdateStatus(
 		return errors.Wrapf(err, "error retrieving event %q from store", eventID)
 	}
 
+	return w.updateStatus(ctx, event, status)
+}
+
+func (w *workersService) Cleanup(
+	ctx context.Context,
+	eventID string,
+) error {
+	if err := w.authorize(ctx, RoleObserver, ""); err != nil {
+		return err
+	}
+
+	event, err := w.eventsStore.Get(ctx, eventID)
+	if err != nil {
+		return errors.Wrapf(err, "error retrieving event %q from store", eventID)
+	}
+
+	return w.cleanup(ctx, event)
+}
+
+func (w *workersService) Timeout(
+	ctx context.Context,
+	eventID string,
+) error {
+	if err := w.authorize(ctx, RoleObserver, ""); err != nil {
+		return err
+	}
+
+	event, err := w.eventsStore.Get(ctx, eventID)
+	if err != nil {
+		return errors.Wrapf(err, "error retrieving event %q from store", eventID)
+	}
+
+	now := time.Now()
+	status := event.Worker.Status
+	status.Phase = WorkerPhaseTimedOut
+	status.Ended = &now
+	if err = w.updateStatus(ctx, event, status); err != nil {
+		return errors.Wrapf(
+			err,
+			"error updating status for event %q worker",
+			event.ID,
+		)
+	}
+
+	// Update all applicable worker job statuses with JobPhaseAborted
+	// TODO: go func here?
+	// TODO: How about errors?  Do we want to compile and report only
+	// after all have been processed?
+	for _, job := range event.Worker.Jobs {
+		if job.Status != nil && !job.Status.Phase.IsTerminal() {
+			status := *job.Status
+			status.Phase = JobPhaseAborted
```

Comment:
This might not always be right.

iirc, when we cancel workers, they go to ABORTED if they were RUNNING, but they go to CANCELED if they were PENDING.

We should probably do the same here. Timed out worker == RUNNING jobs ABORTED and PENDING jobs CANCELED.

Reply:
Just gave myself a refresher on how we cancel events in the db and it involves some fairly complex updates.

I'm almost thinking the best thing to do here is add a dedicated timeout operation on the worker store so that we can quickly and efficiently change the status of all associated jobs that are either pending or starting/running.

Reply:
> I'm almost thinking the best thing to do here is add a dedicated timeout operation on the worker store so that we can quickly and efficiently change the status of all associated jobs that are either pending or starting/running.

I'll give this approach a go.  I should look at the `Cancel` and `CancelMany` implementations in events_store for similar logic, right?

Reply:
That sounds like a good starting point.

Reply:
Great call on this one.  Updated in https://github.com/brigadecore/brigade/pull/1419/commits/3162975063ba74beaf30dfaeddb4f60d4005f920; PTAL when convenient.